### PR TITLE
Combined plant hydraulics and variable soil thickness.

### LINKED ISSUE #3944: Energy balance issue and NaN for F20TRC5-CMIP6 compset with variable soil depth
1. Energy balance error for wetland:
 322:  local  column   index =       304438
 322:  global column   index =       629734
 322:  global landunit index =       203950
 322:  global gridcell index =        63096
 322:  gridcell longitude    =   -124.750000000000
 322:  gridcell latitude     =    43.2500000000000
 322:  column   type         =            6
 322:  landunit type         =            6
 322:  ENDRUN:
 322:  ERROR in BalanceCheckMod.F90 at line 704

2. NaN was noticed in output variable TWS by Michael Brunke.

### PR DESCRIPTION
**Pull request by pnlfang**
This change will allow simulations with both plant hydraulics and variable soil thickness options on.
Minor bug fix for NaNs of TWS and wetland energy balance issue when only variable soil thickness is active. 

Fixes #3944

[BFB]

### CODE REVIEW

**Comment by rljacob** [comment]
@pnlfang is the BFB for v2 watercycle cases?

**Comment by rljacob** [comment]
Is this needed for any v2 sims?

**Comment by rljacob** [comment]
@pnlfang please point to the issue describing the bug this PR fixes.  Or create one if necessary.

**Comment by pnlfang** [comment]
@rljacob BFB is for water cycle cases. It's not needed for v2 sims. As for the bug fix: 
1. There is energy balance error for wetlands caused by inconsistent bedrock heat capacity. See https://github.com/E3SM-Project/E3SM/blob/934a7e6585457a22779130e5f090dce140311798/components/elm/src/biogeophys/SoilStateType.F90#L572 and https://github.com/E3SM-Project/E3SM/blob/934a7e6585457a22779130e5f090dce140311798/components/elm/src/biogeophys/SoilTemperatureMod.F90#L1034
2. The code gave total water storage NaNs error because these two lines (https://github.com/E3SM-Project/E3SM/blob/fcaeac651209f1510c6d44d4c64a4bbe9666e79a/components/elm/src/biogeophys/SoilHydrologyMod.F90#L1106, https://github.com/E3SM-Project/E3SM/blob/fcaeac651209f1510c6d44d4c64a4bbe9666e79a/components/elm/src/biogeophys/SoilHydrologyMod.F90#L1189 ) didn't account for variable soil thickness.

**Comment by bishtgautam** [comment]
@pnlfang You need to create a new issue under https://github.com/E3SM-Project/E3SM/issues and then add "Fixes #<new-issue>" in the first comment of this PR. Here is an example: #3875 was an issue that was closed by PR #3874. See the first comment in PR #3874.

File: components/elm/src/biogeophys/PhotosynthesisMod.F90 [hunk 1]
```diff
@@ -1943,8 +1945,10 @@ subroutine PhotosynthesisHydraulicStress ( bounds, fn, filterp, &
          froot_carbon(p) = tlai(p) / slatop(ivt(p))*froot_leaf(ivt(p))
          croot_carbon(p) = tlai(p)/slatop(ivt(p))*stem_leaf(ivt(p)) &
                   *croot_stem(ivt(p))
-
-         do j = 1,nlevsoi
+!         nlevbed = min(nlev2bed(c),nlevsoi)
+         nlevbed = nlev2bed(c)
+!         nlevbed = nlevsoi
+         do j = 1,nlevbed
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/PhotosynthesisMod.F90:1951]: Remove the commented code

File: components/elm/src/biogeophys/PhotosynthesisMod.F90 [hunk 2]
```diff
@@ -3559,15 +3566,19 @@ logical                , intent(out) :: flag            ! tells calling function
          laisha        => canopystate_inst%laisha_patch         , & ! Input:  [real(r8) (:)   ]  shaded leaf area
          htop          => canopystate_inst%htop_patch           , & ! Input:  [real(r8) (:)   ]  patch canopy top (m)
          tsai          => canopystate_inst%tsai_patch           , & ! Input:  [real(r8) (:)   ]  patch canopy one-sided stem area index, no burying by snow
+         nlev2bed      =>    col_pp%nlevbed                , & ! Input:  [integer  (:)   ]  number of layers to bedrock                     
          ivt           => veg_pp%itype                             & ! Input:  [integer  (:)   ]  patch vegetation type
          )
     
+    nlevbed = min(nlevsoi,nlev2bed(c))
+!         nlevbed = nlev2bed(c)
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/PhotosynthesisMod.F90:3574]: Remove the commented code

File: components/elm/src/biogeophys/PhotosynthesisMod.F90 [hunk 3]
```diff
@@ -3725,11 +3737,16 @@ subroutine spacF(p,c,x,f,qflx_sun,qflx_sha, &
          tsai          => canopystate_inst%tsai_patch           , & ! Input:  [real(r8) (:)   ]  patch canopy one-sided stem area index, no burying by snow
          smp           => soilstate_inst%smp_l_col              , & ! Input: [real(r8) (:,:) ]  soil matrix potential [mm]
          qflx_tran_veg => veg_wf%qflx_tran_veg    , & ! Input:  [real(r8) (:)   ]  vegetation transpiration (mm H2O/s) (+ = to atm)
+         nlev2bed           =>    col_pp%nlevbed                           , & ! Input:  [integer  (:)   ]  number of layers to bedrock                     
          z             => col_pp%z                                   & ! Input:  [real(r8) (:,:) ]  layer node depth (m)
          )
+
+    nlevbed = min(nlevsoi,nlev2bed(c))
+!         nlevbed = nlevsoi
+!         nlevbed = nlev2bed(c)
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/PhotosynthesisMod.F90:3746]: Remove the commented code

File: components/elm/src/biogeophys/PhotosynthesisMod.F90 [hunk 4]
```diff
@@ -3813,23 +3831,28 @@ subroutine getvegwp(p, c, t, x, gb_mol, gs_mol_sun, gs_mol_sha, qsatl, qaf, soil
          hk_l          => soilstate_inst%hk_l_col               , & ! Input: [real(r8) (:,:) ]  hydraulic conductivity (mm/s)
          hksat         => soilstate_inst%hksat_col              , & ! Input: [real(r8) (:,:) ]  hydraulic conductivity at saturation (mm H2O /s)
          sucsat        => soilstate_inst%sucsat_col             , & ! Input: [real(r8) (:,:) ]  minimum soil suction (mm)
+         nlev2bed           =>    col_pp%nlevbed                           , & ! Input:  [integer  (:)   ]  number of layers to bedrock                     
          z             => col_pp%z                                   & ! Input: [real(r8) (:,:) ]  layer node depth (m)
          )
+
+    nlevbed = min(nlevsoi,nlev2bed(c))
+!         nlevbed = nlevsoi
+!         nlevbed = nlev2bed(c)
```
Comment (bishtgautam) [inline components/elm/src/biogeophys/PhotosynthesisMod.F90:3840]: Remove the commented code

**Review by bishtgautam — CHANGES_REQUESTED** [review]
Remove the commented code in various files

File: components/elm/src/biogeophys/PhotosynthesisMod.F90 (hunk 4 above)
Comment (pnlfang) [reply components/elm/src/biogeophys/PhotosynthesisMod.F90]: Removed.

**Comment by pnlfang** [comment]
Fixes https://github.com/E3SM-Project/E3SM/issues/3944.

**Comment by rljacob** [comment]
Notes from IG telecon:  may need rebase.  Ready.

**Review by bishtgautam — APPROVED** [review]

**Comment by rljacob** [comment]
@pnlfang you need to figure out why this was non-BFB when tested.

**Comment by bishtgautam** [comment]
Closing this as PR as no work has been done on it to fix it.

**Pull request closed without merging**